_posts/please-welcome-model-binding-formatters-to-servicecomposer.md

```diff
@@ -0,0 +1,156 @@
+    var body = await reader.ReadToEndAsync();
+    var content = JObject.Parse(body);
+
+    //use the content object instance as needed, e.g.:
+    var aString = content?.SelectToken("AString")?.Value<string>();
+}
+```
+
+Too many things are happening simultaneously; we first need to rewind the stream, there might have been other handlers that processed the request body, and thus, the stream position is at the end. We need to know the request encoding; in the above sample, UTF8 is hardcoded, and it's also crucial to leave the stream open when the read operation completes. Finally, we can't assume the content is a json object, parse and use it. If we don't want to deserialize to a POCO object, we need to deal with the JObject API.
+
+## Model binding to the rescue
+
+Starting with version 1.9.0, ServiceComposer supports ASP.Net model binding. Using the above sample, we need to bind two things, the request body and one route value, the ID parameter. We first need a model for the request body:
+
+```csharp 
+class BodyModel
+{
+    public string AString { get; set; }
+    //all the properties that we need
+}
+```
+
+If we were using ASP.Net controllers, that would have been enough. A controller action signature like the following describes the intention in a good enough manner for the binder to understand what to do:
+
+```csharp 
+[HttpPost("sample/{id}")]
+public Task<object> Sample(int id, [FromBody]BodyModel model)
+```
+
+ASP.Net binding logic matches the first argument name with the route template value key, and thanks to the `[FromBody]` attribute, the second argument gets deserialized from the incoming request body. Pretty straightforward.
+
+In ServiceComposer, we're not yet there, even if we're paving the road to provide a similar experience. We need to explain to the binder where the values we want to bind to are coming from; we can do that with an intermediate class:
+
+```csharp
+class RequestModel
+{
+    [FromRoute] public int id { get; set; }
+    [FromBody] public BodyModel Body { get; set; }
+}
+```
+
+The `ReqestModel` class is what glues all the concepts that we need. We have a public property, id, that matches the route value key and is decorated with the `[FromRoute]` attribute to explain to the binder where the value comes from, and public property for the body using binding attributes as well.
+
+> By the way, class names can be whatever you want.
+
+With the two above classes in place, we can now write the following handler:
+
+```csharp
+[HttpPost("sample/{id}")]
+public async Task Handle(HttpRequest request)
+{
+    var requestModel = await request.Bind<RequestModel>();
+    var body = requestModel.Body;
+    var aString = body.AString;
+    var id = requestModel.id;
+
+    //use values as needed
+}
+```
+
+The presented code, thanks to model binding, becomes much more maintainable and easy to understand. Not to mention that it is now compiler safe, being strongly typed.
```

Comment:
Just a suggestion to remove an excess comma:

```suggestion
Thanks to model binding, the presented code becomes much more maintainable and easier to understand. Not to mention that it is now compiler safe, since it's strongly typed.
```